client/gogeta_client.go

```diff
@@ -41,39 +55,121 @@ func (client *GogetaClient) Start() {
+	notifications := publisher.SimpleNotification{
+		Application: &amazonSNS,
+		Log:         log,
+	}
+
+	// Setup client
 	client.Log = log
-	client.SCM = scm
 	client.Storage = store
+	client.Queue = amazonQueue
+	client.Notifications = &notifications
 }
 
-// GetSourceCode requests messages from the queues and gathers source code
-// if the messages are not empty
-func (client *GogetaClient) GetSourceCode() *sourcesystem.SourceRepository {
+// RunGogetaClient will run the complete gogeta scm system
+func (client *Gogeta) RunGogetaClient() *sourcesystem.SourceRepository {
 	repo := sourcesystem.SourceRepository{}
+	client.queueMessages()
+	if len(client.Data) <= 0 {
+		return &repo
+	}
+
+	client.setVersionControl()
+	if client.SCM == nil {
+		return &repo
+	}
+
+	client.downloadSource(&repo)
+	if repo.SourceLocation == "" {
+		return &repo
+	}
+	client.archiveRepo(&repo)
+	client.notifyMrRobot(&repo)
+	return &repo
+}
+
+func (client *Gogeta) queueMessages() {
 	messages, err := client.Queue.GetQueueMessages()
 	if err != nil {
 		client.Log.Error(err.Error())
 	}
+	client.Data = messages
+}
 
-	if len(messages) <= 0 {
-		return &repo
+func (client *Gogeta) setVersionControl() {
+	dataType := client.Data[0].Type
+	switch dataType {
+	case "GIT":
+		scm := &sourcesystem.SystemSCM{}
+		scm.VersionControl = &sourcesystem.GitVersionControl{}
+		client.SCM = scm
+		return
+	default:
+		client.Log.Info("SCM Type not found: " + dataType)
+		return
 	}
+}
+
+func (client *Gogeta) downloadSource(repo *sourcesystem.SourceRepository) {
+	message := client.Data[0]
+	project := message.Project
+	origin := message.Repo
 
-	project := messages[0].Proj
-	origin := messages[0].Repo
 	if project == "" || origin == "" {
-		return &repo
+		return
 	}
 
 	repo.ProjectName = project
 	repo.SourceOrigin = origin
-	if err := client.SCM.AddSource(&repo); err != nil {
+
+	if err := client.SCM.AddSource(repo); err != nil {
 		client.Log.Error(err.Error())
 	}
-	return &repo
+}
+
+func (client *Gogeta) archiveRepo(repo *sourcesystem.SourceRepository) {
+	archive := path.Join(os.Getenv("GOPATH"), "/repos/", repo.ProjectName+".zip")
+	storageData := storehouse.StorageData{
+		Source: repo.SourceLocation,
+		Target: archive,
+	}
+	if err := client.Storage.StoreFiles(&storageData); err != nil {
+		client.Log.Error(err.Error())
+	}
+	client.Data[0].ArchivePath = repo.ProjectName + ".zip"
+}
+
+func (client *Gogeta) notifyMrRobot(repo *sourcesystem.SourceRepository) {
+	data := client.Data[0]
+	message := MrRobotMessage{
+		ArchivePath:    data.ArchivePath,
+		Project:        data.Project,
+		EngineName:     data.EngineName,
+		EnginePlatform: data.EnginePlatform,
+		EngineVersion:  data.EngineVersion,
+		BuildrID:       data.BuildrID,
+		BuildID:        data.BuildID,
+	}
+	jsonMessage, err := json.Marshal(message)
+	if err != nil {
+		client.Log.Error(err.Error())
+		return
+	}
+	notification := publisher.Message{
+		JSON:     jsonMessage,
+		Subject:  "Buildr Request",
```

Comment:
Don't think it matters does it? I'm not sure I ever actually use the message topic.